### DISCUSSION TITLE
Syntax-highlighting for Nano

### DISCUSSION THREAD

**Discussion by brainary** [jotaen/xit#21] [Show & tell]
First attempt at syntax-highlighting of **[x]it!** for Nano. (Actually first attempt at all at highlighting for nano)

File: xit.nanorc
https://gist.github.com/brainary/a5792467e931e4466fdb79b8b18486ae

Include it with, in your **.nanorc**
`include "xit.nanorc"`

Doesn't seem like **nano** allows look ahead, so this uses layering, so  the order in the nanorc file is significant.

Bellow is the file from the main site shown in **nano 5.6.1**
![image](https://user-images.githubusercontent.com/5373426/161441799-460a5298-97fe-46d4-b2e0-4d6e83d6ad71.png)

**Comment by jotaen** [comment]
🎉  That looks great! Nano is my favourite CLI-based editor. I also have never written any plugin for it, and I’m amazed to see how slim it is.

Let me know if (or when) you would like to share this on [the website](https://xit.jotaen.net/). It’s just an offer, so don’t feel pushed. If you don’t want to maintain a proper repository, we could also just link this discussion for other nano-users to have a starting point.

Maybe you have seen it already, but I posted something [here about testing](https://github.com/jotaen/xit/discussions/6) the highlighting. Maybe that’s helpful for you, since the demo data on the website doesn’t cover any edge-cases.

**Reply by brainary** [reply]
Share it whenever you want, it's not perfect. And it might never be, due to the way highlighting in nano works.
Yeah, I tried to make a test suit to find edge-cases (se other comment), some are hard/impossible to fix though.

**Comment by brainary** [comment]
Testsuit and Result of the nano syntax highlighting.

Problem areas; 
- NOT tags
- NOT priorities (tried to add a "top" layer to force invalid spacing to show an error, but it didn't really work)
- NOT anything (specifically Headline under item, hard without look-ahead.

Test;
[test.xit.txt](https://github.com/jotaen/xit/files/8405278/test.xit.txt)

Result (2 images);
![image](https://user-images.githubusercontent.com/5373426/161443654-20197ad6-aab8-4adc-a449-8af8c8f62a7d.png)
![image](https://user-images.githubusercontent.com/5373426/161443667-cde64800-c333-4a04-a54f-82bb5402a36e.png)

**Reply by jotaen** [reply]
I’ve linked the gist from the website now, so that other nano users are able to discover it.

The test file looks actually pretty good to me. It’s probably not possible to get all the rules right in every editor. E.g. if you consider something like [“redundant due dates should be disregarded”](https://github.com/jotaen/xit/blob/main/Specification.md#description). But I think that’s totally okay.

I also want to work on [improving the test file](https://github.com/jotaen/xit/issues/22).

> NOT tags

I think getting the `a#tag` one right is not possible without regex lookbehind. The other `NOT Tags` ones look correct, though. ([If the closing quote is missing, the value should be treated as absent](https://github.com/jotaen/xit/blob/main/Specification.md#tag).)

> NOT priorities

In my book, the invalid highlighting is just a “bonus”, because users should already notice that the syntax is incorrect if it’s not highlighted (which it seems to do in your screenshot).

Just out of curiosity: do you think a feature like [nested items](https://github.com/jotaen/xit/discussions/2) would be doable in nano? (I mean a somewhat correct syntax highlighting for it.)

**Reply by brainary** [reply]
Maybe.

Multi-line items coloring is currently cheating in the highlighting. (Multiple lines of the same type are selected in one go)

Changing the selectors to handle indented Items should make it harder to mark them as an error, but like you said that might not really be needed, as for a editor as simple as nano it might just be more of a guide.

Nano 5+ does support formatting and linting, so a simpler syntax-highlighting and formatting might be a way to fix stuff.

**Comment by jotaen** [comment]
Just fyi, for improved test data: https://github.com/jotaen/xit/issues/22
